### PR TITLE
Fix #1433: Restructure the Logger 

### LINKED ISSUE #1433: Restructure the Logger
With the way we have loggers right now, app/domain files would have to fetch multiple types of loggers for different types of logging. We want to refactor it so that atleast app/domain modules can use a single logger: Oppia Logger.
This restructuring will have multiple parts to it:

- [x] Rename Logger.kt -> ConsoleLogger.kt since that's what it does (Data/Utility module files will use this file directly)

- [x] Create an OppiaLogger.kt file in the domain module. All App/Domain module files will use this logger

- [x] Oppia Logger will route the log command to the lower lying logger:
       - logException routes to ExceptionLogger (or exceptionController)
       - logTransition/logClick routes to the AnalyticsController
       - .v/.d etc commands route to the ConsoleLogger

### PR DESCRIPTION
**Pull request by Aditiwebd**
## Explanation
Restructure the different loggers into one file i.e OppiaLogger.kt

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by Aditiwebd** [comment]
I know it's not all the work. But I just wanted to confirm If I am on the right path. @Sarthak2601

**Comment by Sarthak2601** [comment]
> I know it's not all the work. But I just wanted to confirm If I am on the right path. @Sarthak2601

Yes, you're on the right track. However, the tests are failing so please take a look at that as well.

**Comment by Aditiwebd** [comment]
PTAL @Sarthak2601

**Comment by Aditiwebd** [comment]
Could you help me with what went wrong @Sarthak2601 The ApplicationComponent file has an error of Dependency Cycle\Dagger.

File: domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt [hunk 1]
```diff
@@ -114,4 +118,62 @@ class OppiaLogger @Inject constructor(
...
+  /**Logs non-fatal exception. See [ExceptionsController.logNonFatalException] for more context.*/
+  fun logNonFatalException(exception: Exception, timestampInMillis: Long) {
+    exceptionController.logNonFatalException(exception, timestampInMillis)
+  }
+
+  /**Logs exception. See [ExceptionsController.logNonFatalException] for more context.*/
+  fun logFatalException(exception: Exception, timestampInMillis: Long) {
+    exceptionController.logFatalException(exception, timestampInMillis)
+  }
+  fun v(tag: String, msg: String) {
```
Comment (Sarthak2601) [inline domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt:133]: Please add a blank line and a Kdoc comment describing `fun v` here.

File: domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt [hunk 2]
```diff
@@ -114,4 +118,62 @@ class OppiaLogger @Inject constructor(
       )
       .build()
   }
+
+  // Exception Controller
```
Comment (Sarthak2601) [inline domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt]: No need to add `//Exception Controller`.

**Review by Sarthak2601 — CHANGES_REQUESTED** [review]
Thanks @Aditiwebd! Few thoughts.

**Comment by Sarthak2601** [comment]
> Could you help me with what went wrong @Sarthak2601 The ApplicationComponent file has an error of Dependency Cycle\Dagger.

Aah, it's happening because you're using `OppiaLogger` in `AnalyticsController` and `ExceptionsController` (for console logging) and then using these two controllers in oppiaLogger (for event and exception logging). I think keeping the console logging in these two files via the consoleLogger will solve this issue.

**Review by vinitamurthi — COMMENTED** [review]
I see a lot more files that use the consoleLogger, for example: https://github.com/oppia/oppia-android/blob/abc28622c9b891078452502e36b2cee175fb73e0/app/src/main/java/org/oppia/android/app/topic/conceptcard/ConceptCardViewModel.kt

Here's the entire list: https://github.com/oppia/oppia-android/search?p=2&q=consoleLogger

Can we make sure we use oppiaLogger in place of consoleLogger in all the files that are within the app/ or domain/ folder?

**Comment by Aditiwebd** [comment]
Oh, Thanku @Sarthak2601. I only changed the domain files I will update the others also. Thanku

**Comment by Sarthak2601** [comment]
@Aditiwebd please take care of the test cases. They are generally lint and dagger injection related failures.

File: domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt (hunk 1 above)
Comment (Aditiwebd) [reply domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt:133]: Done

File: domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt (hunk 2 above)
Comment (Aditiwebd) [reply domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt]: Done

File: domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt [hunk 3]
```diff
@@ -114,4 +118,64 @@ class OppiaLogger @Inject constructor(
       )
       .build()
   }
+
+  /**Logs non-fatal exception. See [ExceptionsController.logNonFatalException] for more context.*/
```
Comment (Sarthak2601) [inline domain/src/main/java/org/oppia/android/domain/oppialogger/OppiaLogger.kt:122]: I think it'd be better if we move the context creation related functions towards the end of this file.

File: domain/src/test/java/org/oppia/android/domain/oppialogger/OppiaLoggerTest.kt [hunk 4]
```diff
@@ -120,6 +126,28 @@ class OppiaLoggerTest {
...
+    val exceptionThrown = Exception("TEST MESSAGE", Throwable())
+    oppiaLogger.logFatalException(
+      exceptionThrown, TEST_TIMESTAMP_IN_MILLIS_ONE
+    )
+
+    val exceptionLogged = fakeExceptionLogger.getMostRecentException()
+
+    assertThat(exceptionLogged).isEqualTo(exceptionThrown)
+  }
+
```
Comment (Sarthak2601) [inline domain/src/test/java/org/oppia/android/domain/oppialogger/OppiaLoggerTest.kt:151]: I think there are more tests that can be added here for the exceptions like when there is no network connectivity and they are stored into the cache store. Refer to the `ExceptionControllerTests` for these.

**Review by Sarthak2601 — CHANGES_REQUESTED** [review]
Nice work @Aditiwebd ! Left some thoughts.

**Comment by Sarthak2601** [comment]
@vinitamurthi I'd like your thoughts on the following - 
1. Not all files in the domain layer will be able to use the OppiaLogger as if we use it to write console logs from the Analytics and Exceptions Controllers, the issue of circular dependency arises. So is it okay to leave the console logging in these two with the consoleLogger itself? 
2. We currently don't have tests written for the consoleLogger in utils, should we create them in the OppiaLogger tests or leave it as is?

File: domain/src/test/java/org/oppia/android/domain/oppialogger/OppiaLoggerTest.kt (hunk 4 above)
Comment (Aditiwebd) [reply domain/src/test/java/org/oppia/android/domain/oppialogger/OppiaLoggerTest.kt:151]: The other test cases required some ```exceptionController``` functions. Should I put it like that? @Sarthak2601 for ex: 
Capture
 The ```getExceptionLogStore``` function is in ```exceptionsController```. I tried adding the function in ```oppiaLogger``` just to check then the ```LiveData()``` had error.

**Comment by vinitamurthi** [comment]
> * Not all files in the domain layer will be able to use the OppiaLogger as if we use it to write console logs from the Analytics and Exceptions Controllers, the issue of circular dependency arises. So is it okay to leave the console logging in these two with the consoleLogger itself?

I think anything in the oppiaLogger folder itself will use consolelogger, not any other file so yes Analytics and Exception Controllers do not need to use oppialogger

> * We currently don't have tests written for the consoleLogger in utils, should we create them in the OppiaLogger tests or leave it as is?

Hmm, I think we can add tests for them , it should be pretty straightforward!

**Comment by vinitamurthi** [comment]
I will review this once again after tests are added, but current changes LGTM

File: domain/src/test/java/org/oppia/android/domain/oppialogger/OppiaLoggerTest.kt (hunk 4 above)
Comment (Sarthak2601) [reply domain/src/test/java/org/oppia/android/domain/oppialogger/OppiaLoggerTest.kt:151]: I think you can use these functions in the tests without adding them to the OppiaLogger.

**Comment by Sarthak2601** [comment]
> > * Not all files in the domain layer will be able to use the OppiaLogger as if we use it to write console logs from the Analytics and Exceptions Controllers, the issue of circular dependency arises. So is it okay to leave the console logging in these two with the consoleLogger itself?
> 
> I think anything in the oppiaLogger folder itself will use consolelogger, not any other file so yes Analytics and Exception Controllers do not need to use oppialogger
> 
> > * We currently don't have tests written for the consoleLogger in utils, should we create them in the OppiaLogger tests or leave it as is?
> 
> Hmm, I think we can add tests for them , it should be pretty straightforward!

Okay, thanks @vinitamurthi. 
@Aditiwebd Please add test cases for the consoleLogger functions as well in the OppiaLogger tests.

**Comment by Aditiwebd** [comment]
I will be creating a separate PR for the testing of console logger. I have to implement the fake one first. I will need some help on it if either of you have some time can we have a meeting on how to do it @Sarthak2601 @vinitamurthi ? Thankyou

**Comment by Aditiwebd** [comment]
ALso I am having some problems on the exceptionsController tests in oppiaLoggerTests. I have added two more tests and 
Capture
I get this error. I tried adding the getExceptionLogStore() in oppiaLogger but it still gives the same error.

**Comment by anandwana001** [comment]
@Sarthak2601 Could you please look into this?

**Comment by Sarthak2601** [comment]
> @Sarthak2601 Could you please look into this?

Okay, I'll set up a meeting with the contributor and help her out.

**Comment by anandwana001** [comment]
@Sarthak2601 @Aditiwebd Any update on this? 
I see conflicts increasing as this PR covers a large number of files.

**Comment by Luffy18346** [comment]
@Sarthak2601 @Aditiwebd Any update on this?

**Comment by Aditiwebd** [comment]
So sorry for the delay. I had the meeting with @Sarthak2601 I will complete it in this week only. Thank You.

**Comment by Luffy18346** [comment]
> So sorry for the delay. I had the meeting with @Sarthak2601 I will complete it in this week only. Thank You.

Ok. Please resolve conflicts when you work on this.

**Comment by Aditiwebd** [comment]
I don't want to merge it. I was just solving the conflicts. I think I clicked something wrong.

**Comment by anandwana001** [comment]
@Aditiwebd any update on this?

**Comment by rt4914** [comment]
Closing this PR because of no activity. Thanks @Luffy18346 for the update.

**Pull request closed without merging**